cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -249,7 +249,10 @@ class ObjectInputFile : public io::RandomAccessFile {
       if (IsNotFound(outcome.GetError())) {
         return PathNotFound(path_);
       } else {
-        return ErrorToStatus(outcome.GetError());
+        return ErrorToStatus(
+            std::forward_as_tuple("When reading information for key '", path_.key,
+                                  "' in bucket '", path_.bucket, "': "),
+            outcome.GetError());
```

Comment:
@fsaintjacques What do you think about this approach?

Reply:
So from what I understand, the `std::tuple` gimmick is because otherwise Arg&... needs to be the last argument. The usage is odd. May I suggest to refactor `ErrorToStatus` to a single class, e.g. `ToStatus` where the constructor takes a single Arg&... and you provide to "build" functions for both error and outcome, e.g.

```
template <Arg&...>
class ToStatus {
   public:
   ToStatus(Args&.... args) : prefix_(util::StringBuilder(std::forward(args)) {}

   template<typename ErrorType>
   Status With(const Aws::Client::AWSError<ErrorType>& error)...

   template<typename ErrorType>
   Status With(const Aws::Utils::Outcome<Result, Error>& outcome)...

   private:
      std::string prefix_;
}
```

The usage would look like 

```
ToStatus("When reading information for key '", path_.key, "' in bucket '", path_.bucket, "': ").With(outcome);`
```

or, inverse the argument order and make error/outcome first (even if it's odd for the message format), that way you can get rid of tuple.

Reply:
The annoying thing with calling `StringBuilder` is that it will format arguments even when ok. This is not really a problem here, but it makes the design less generic.

It would be nice if we could create a "formatting object" that's both generic and lazy, and that we could pass as an argument to Status...